upi/openstack/library/os_create_svc_subnet.py

```diff
@@ -0,0 +1,123 @@
+     description:
+        - Network this subnet belongs to
+     required: true
+   cidr:
+     description:
+       - Subnet range in CIDR notation
+     required: true
+   openstack_svc_network:
+     description:
+        -  CIDR of network to allocate IPs for OpenStack
+           Octavia's Amphora VMs. Note that the value of this
+           option must be at least two times the size of the
+           cidr option, as octavia uses two IPs from this network
+           for each loadbalancer, one given by OpenShift and other
+           for VRRP connections.
+     required: true
+   ip_version:
+     description:
+       - IP version
+     required: true
+   enable_dhcp:
+     description:
+        - Enable DHCP
+     required: true
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.openstack import openstack_full_argument_spec, openstack_module_kwargs, openstack_cloud_from_module
+
+import ipaddress
+
+def main():
+    ''' Main module function '''
+    argument_spec = openstack_full_argument_spec(
+        subnet_name=dict(type='str', required=True),
+        network_name=dict(type='str', required=True),
+        cidr=dict(type='str', required=True),
+        openstack_svc_network=dict(type='str', required=True),
+        ip_version=dict(type='int', required=True),
+        enable_dhcp=dict(type=bool, required=True))
+
+    module_kwargs = openstack_module_kwargs()
+    module = AnsibleModule(argument_spec, **module_kwargs)
+
+    subnet_name = module.params['subnet_name']
+    network_name = module.params['network_name']
+    cidr = module.params['cidr']
+    openstack_cidr = module.params['openstack_svc_network']
+    ip_version = module.params['ip_version']
+    enable_dhcp = module.params['enable_dhcp']
+
+    svc_network = ipaddress.ip_network(cidr)
+    openstack_svc_cidr = ipaddress.ip_network(openstack_cidr)
+
+    # Parts of openstack_svc_cidr not overlapping with svc_network
+    allocation_ranges = []
+    # If the last IPs are different the first
+    # part of the openstack network is already allocated
+    if svc_network[-1] != openstack_svc_cidr[-1]:
+        allocation_ranges.extend([{
+            "start": str(svc_network[-1]+1),
+            "end": str(openstack_svc_cidr[-2])}])
+
+    # If the first IPs are different the last part
+    # of the openstack network is already allocated
+    if svc_network[0] != openstack_svc_cidr[0]:
+        allocation_ranges.extend([{
+            "start": str(openstack_svc_cidr[0]+1),
+            "end": str(svc_network[0]-1)}])
+
+    if allocation_ranges:
+        gateway_ip = str(allocation_ranges[-1]["end"])
+        allocation_ranges[-1]["end"] = str(ipaddress.IPv4Network(gateway_ip)[0]-1)
+
+    sdk, cloud = openstack_cloud_from_module(module)
+    try:
+        subnet = cloud.network.find_subnet(name_or_id=subnet_name)
+        network = cloud.network.find_network(name_or_id=network_name)
+        if not subnet and network and allocation_ranges:
```

Comment:
should this have an else too? indicating there is no network or allocation_range? It is not the same that the subnet already exits than not creating it since the network is not there.

Similarly, what would happen if the network exits but the range are not the ones provided?

Reply:
I will account for those cases.